### PR TITLE
[AMDGPU][True16][CodeGen] support v_mov_b16 and v_swap_b16 in true16 format

### PR DESCRIPTION
**Pull request by broxigarchen**
support v_swap_b16 in true16 format.
update tableGen pattern and folding for v_mov_b16.

### CODE REVIEW

**Comment by broxigarchen** [comment]
Hi @Sisyph @kosarev @jayfoad @rampitec  can you help to review this PR? Thanks!

File: llvm/lib/Target/AMDGPU/SIFoldOperands.cpp [hunk 1]
```diff
@@ -1460,7 +1460,15 @@ bool SIFoldOperands::tryFoldFoldableCopy(
     return false;
   }
 
-  MachineOperand &OpToFold = MI.getOperand(1);
+  MachineOperand *OpToFoldPtr;
+  if (MI.getOpcode() == AMDGPU::V_MOV_B16_t16_e64) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:1464]: I don't expect to need special casing here. Should this be filtered in isFoldableCopy? I thought we had filtering for regular v_mov_b32_e64 already somewhere here

Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:1464]: Hi Matt. v_mov_b32_e64 only has 1 source operand so we don't the filtering.  
V_MOV_B16_t16_e64 has three input operands and  I think it is only instruction in the FoldableCopy list that has more than 1 operand:
`inoperandlist: src_modifers, src0, op_sel`
so it need some special handling here.

I think we could move the filtering to isFoldableCopy though, but we need still need the special handling to select correct operand. What do you think?

File: llvm/lib/Target/AMDGPU/SIInstructions.td [hunk 2]
```diff
@@ -2192,20 +2192,6 @@ foreach pred = [NotHasTrue16BitInsts, UseFakeTrue16Insts] in {
   }
 }
 
-let True16Predicate = UseRealTrue16Insts in {
```
Comment (Sisyph) [inline llvm/lib/Target/AMDGPU/SIInstructions.td:2195]: Why are you removing these patterns?

Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/SIInstructions.td]: I just realized these are patterns merged last week. Restored

**Review by Sisyph — APPROVED** [review]
LGTM but please wait for @arsenm 's comment resolution

**Review by arsenm — APPROVED** [review]

**Comment by broxigarchen** [comment]
Hi I think this PR is ready for merge.

Can someone help to merge this PR since I don't have write access? Thanks!

**Comment by hanhanW** [comment]
Hi there, I bisected to this change in a downstream compiler (https://github.com/iree-org/iree) crash on gfx1100 target. The error is `error: Illegal instruction detected: Operand has incorrect register class.`.

I'm going to revert it locally in our project. I don't know how to generate a repro for you at this moment, but I'll try to do that. I'm asking questions about generating a repro in https://github.com/iree-org/iree/pull/18177#issuecomment-2278892368.

**Comment by broxigarchen** [comment]
> Hi there, I bisected to this change in a downstream compiler (https://github.com/iree-org/iree) crash on gfx1100 target. The error is `error: Illegal instruction detected: Operand has incorrect register class.`.
> 
> I'm going to revert it locally in our project. I don't know how to generate a repro for you at this moment, but I'll try to do that. I'm asking questions about generating a repro in [iree-org/iree#18177 (comment)](https://github.com/iree-org/iree/pull/18177#issuecomment-2278892368).

Ran a quick repro. These are the error info:

error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.
V_CMP_EQ_U16_t16_e32 0, killed $vgpr254, implicit-def $vcc, implicit $exec
error: Illegal instruction detected: Operand has incorrect register class.

**Comment by broxigarchen** [comment]
> Hi there, I bisected to this change in a downstream compiler (https://github.com/iree-org/iree) crash on gfx1100 target. The error is `error: Illegal instruction detected: Operand has incorrect register class.`.
> 
> I'm going to revert it locally in our project. I don't know how to generate a repro for you at this moment, but I'll try to do that. I'm asking questions about generating a repro in [iree-org/iree#18177 (comment)](https://github.com/iree-org/iree/pull/18177#issuecomment-2278892368).

Hi @hanhanW Thanks for the notifiication for this error. Just want to double check with you that you are not actively testing anything related with true16 instructions and you did not use +real-true16 in the compile command, is that right?

The error is caused by t16 format instruction being selected while the 128-256 vgpr are mistakenly used. These t16 instructions should not be selected in default/fake16 flow.

**Pull request merged**